### PR TITLE
client-go/util/consistencydetector: improve validation of list parameters (RV, ListOptions)

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
xref: https://github.com/kubernetes/kubernetes/pull/125159

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go [hunk 1]
```diff
@@ -79,6 +83,22 @@ func CheckDataConsistency[T runtime.Object, U any](ctx context.Context, identity
...
+func areParametersValidForListCall(resourceVersion string, options metav1.ListOptions) bool {
+	// since we are setting ResourceVersionMatch to metav1.ResourceVersionMatchExact
+	// we need to make sure that the continuation hasn't been set
+	// https://github.com/kubernetes/kubernetes/blob/be4afb9ef90b19ccb6f7e595cbdb247e088b2347/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go#L38
+	hasContinuation := len(options.Continue) > 0
+
+	// since we are setting ResourceVersionMatch to metav1.ResourceVersionMatchExact
+	// we need to make sure that the RV is valid because the validation code forbids RV == "0"
+	// https://github.com/kubernetes/kubernetes/blob/be4afb9ef90b19ccb6f7e595cbdb247e088b2347/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go#L44
+	validResourceVersion := resourceVersion != "0"
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:97]: not sure about this check.

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go [hunk 2]
```diff
@@ -79,6 +83,22 @@ func CheckDataConsistency[T runtime.Object, U any](ctx context.Context, identity
...
+
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:98]: we could also check `options.ResourceVersionMatch`, if set to `metav1.ResourceVersionMatchExact` we could skip.

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go [hunk 3]
```diff
@@ -41,6 +41,10 @@ type ListFunc[T runtime.Object] func(ctx context.Context, options metav1.ListOpt
 // we cannot manipulate the environmental variable because
 // it will affect other tests in this package.
 func CheckDataConsistency[T runtime.Object, U any](ctx context.Context, identity string, lastSyncedResourceVersion string, listFn ListFunc[T], listOptions metav1.ListOptions, retrieveItemsFn RetrieveItemsFunc[U]) {
+	if !areParametersValidForListCall(lastSyncedResourceVersion, listOptions) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:44]: the name of the function is misleading - the calls are valid for list call.

They may not be valid for our update explicitly setting RV and RVMatch

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:97]: We will update it in line 49 anyway - this should be removed.

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:98]: This is an optimization, we cna get back to it later.

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go (hunk 1 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:97]: line 49 will be updated to the same value passed to this function.

File: staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go (hunk 3 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:44]: Note that the params (lastSyncedResourceVersion, listOptions)) are taken without any modifications to from an additional LIST call. 

We could rename to `areParametersValidToFormAdditionalListCall` or `canFormAdditionalListCall`, wdyt?

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go:44]: canFormAdditionalListCall sounds fine

The alternative would be to first update the listOptions by setting RV/RVM and then call ValidateListOptions.
The drawback then is that we may be not handling something without knowing that...

Comment (p0lyn0mial) [reply staging/src/k8s.io/client-go/util/consistencydetector/data_consistency_detector.go]: I prefer not to modify the options until we know are valid. Updated the PR, PTAL.

**Pull request merged**